gramineproject/gramine
# [packaging] add Alpine Linux build scripts

## Description of the changes
This PR adds build scripts with patches needed to run Gramine on Alpine Linux.

## Features
Everything works*

*YMMV, some test might still error you out since they usually have to assume a glibc-based host, SGX works only through Docker (for now), glibc apps don't work (and are not going to work; we should respect Alpine community's position on mixing glibc with musl; it is possible to build Gramine with glibc, but it's way more hacky than just running a different distro through chroot or Docker)

## How to test this PR?
NOTE: in order to use SGX, you have to run Docker on a host supported by SGX SDK (such as Ubuntu or Alma) and mount its SGX device to Docker, this changes only if PSW/DCAP gets officially released for musl-based distros.
- create a Docker instance of Alpine 3.18/edge with following flags: `--cap-add SYS_ADMIN --cap-add SYS_PTRACE --security-opt seccomp=unconfined --device=/dev/sgx_enclave`
- add Alpine's main, community and testing repos to /etc/apk/repositories
- install doas, wget, alpine-sdk and perl-utils (and optionally git) through apk
- create a new user and assign it to the `wheel` and the `abuild` group
- remove -Wl,-z,pack-relative-relocs from /usr/share/abuild/default.conf
- pull the /alpine/ folder from this PR
- run `abuild-keygen -a -i`
- run `abuild -r checksum all` - this will compile Gramine from the main branch
- you might see an error `gramine*: ld-linux-x86-64.so.2: path not found`, ignore it
- packages got compiled to `~/packages/`, cd to that directory and install them using `apk`
For tests:
- compile Gramine repo and generate SGX privkey
- modify the example manifest.template from `gramine.runtimedir()` to `gramine.runtimedir(libc='musl')`

For bare metal running Alpine, you should set `-Dsgx=disabled` in the APKBUILD file.

Signed-off-by: Amie Raine <amie@invisiblethingslab.com>


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/1421)
